### PR TITLE
Inclusion of Signal Shape Simulation for the Phase2 Tracker

### PR DESCRIPTION
**Pull request by suchandradutta**
#### PR description:
Inclusion of  Time dependence of Signal shape
 - Time dependent Signal Shape correction added for SSDigitizerAlgorithm 

 - Pixel Timewalk correction added for PixelDigitizer Algorithm

 - DigitizerUtility::Amplitude class updated to use newly introduced SimHitInfo class where Time of   Flight information of SimHit is stored so that it can be accessed at the 
Phase2DigitizerAlgorithm::digitize step

  - Configuration parameters are set to switch off the Signal Shape Simulation. So output should be identical with the current code

  -  Integer and float literals are defined as constant. For example, we are now using
     constexpr double c_cm_ns = CLHEP::c_light * CLHEP::ns / CLHEP::cm;
     constexpr double c_inv = 1.0 / c_cm_ns;
      .....
      float corr_time = hit.tof() - pixdet->surface().toGlobal(hit.localPosition()).mag() * c_inv;
      instead of
      float corr_time = hit.tof() - pixdet->surface().toGlobal(hit.localPosition()).mag() / 30.;
     in SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc and in other    places. Minor changes observed in DQM histograms due to such improved accuracy in unit conversions.

  - Does not depend on other PRs or externals

#### PR validation:

Code tested with Particle gun Muon + 200 PU (RelVal) events and the results can be found here
https://indico.cern.ch/event/879031/contributions/3823579/attachments/2018881/3376562/SignalShape_validation_14042020.pdf 

code-checks    done
code-format     done

@emiglior @skinnari @subirsarkar

### CODE REVIEW

**Comment by boudoul** [comment]
Dear @suchandradutta , thanks for the PR. 
Could you please make the title more explicit by identifying which subdetectors are involved (otherwise we cannot guess is this is for ECAL , HGCAL, or inner Pixel and outer Tracker..) - Thanks - Presumably 1110 x 05042020  is also your internal notation from your branch, but not needed in a PR title (which will appear in the release notes) . Thank you.

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc [hunk 1]
```diff
@@ -22,7 +22,6 @@
 #include "SimTracker/SiPhase2Digitizer/plugins/PSSDigitizerAlgorithm.h"
 #include "SimTracker/SiPhase2Digitizer/plugins/PSPDigitizerAlgorithm.h"
 #include "SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.h"
-#include "SimTracker/SiPhase2Digitizer/plugins/Pixel3DDigitizerAlgorithm.h"
```
Comment (mmusich) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc:25]: why the 3D digitizer algorithm changes are being removed?

File: SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py [hunk 2]
```diff
@@ -57,48 +57,29 @@
       EfficiencyFactors_Barrel = cms.vdouble(0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999 ),
       EfficiencyFactors_Endcap = cms.vdouble(0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 0.999, 
       0.999, 0.999 ),#Efficiencies kept as Side2Disk1,Side1Disk1 and so on
-      CellsToKill = cms.VPSet()
-    ),
-#Pixel-3D Digitizer Algorithm
-    Pixel3DDigitizerAlgorithm = cms.PSet(
```
Comment (mmusich) [inline SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py:63]: @suchandradutta why are you removing the configuration for the 3D pixel digitizer algorithm?
@duartej

Comment (duartej) [reply SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py:63]: Thanks @mmusich for noticing this!! 
I saw more removals related with the pixel 3D  in this PR, could you please revert this changes @suchandradutta ?

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc (hunk 1 above)
Comment (duartej) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc:25]: @suchandradutta could you please revert this change?

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc [hunk 3]
```diff
@@ -91,7 +90,6 @@ namespace cms {
     }
     // creating algorithm objects and pushing them into the map
     algomap_[AlgorithmType::InnerPixel] = std::make_unique<PixelDigitizerAlgorithm>(iConfig);
-    algomap_[AlgorithmType::InnerPixel3D] = std::make_unique<Pixel3DDigitizerAlgorithm>(iConfig);
```
Comment (duartej) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc:94]: @suchandradutta could you please revert this change?

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc [hunk 4]
```diff
@@ -250,12 +248,6 @@ namespace cms {
       case TrackerGeometry::ModuleType::Ph2PXF:
         algotype = AlgorithmType::InnerPixel;
         break;
-      case TrackerGeometry::ModuleType::Ph2PXB3D:
```
Comment (duartej) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc:253]: From here up to L258, could you please revert the removal @suchandradutta ?

File: SimTracker/SiPhase2Digitizer/test/BuildFile.xml [hunk 5]
```diff
@@ -19,7 +19,3 @@
   <flags   EDM_PLUGIN="1"/>
   <flags CXXFLAGS="-g3"/>
 </library>
-<library   file="PixelTestBeamValidation.cc" name="PixelTestBeamValidation">
```
Comment (duartej) [inline SimTracker/SiPhase2Digitizer/test/BuildFile.xml:22]: @suchandradutta , could you please revert this validator? As we agreed, the validator will remain in this package until the code is fully validate, afterwards it will be decided if the validator should move to the proper package (Validation) or just discarded.

File: SimTracker/SiPhase2Digitizer/test/BuildFile.xml [hunk 6]
```diff
@@ -19,7 +19,3 @@
...
-  <flags   EDM_PLUGIN="1"/>
-  <flags CXXFLAGS="-g3"/>
-</library>
```
Comment (mmusich) [inline SimTracker/SiPhase2Digitizer/test/BuildFile.xml:25]: Also this should not be removed.

**Review by duartej — COMMENTED** [review]
@suchandradutta , also I would like to know why you remove all pixel 3D related stuff?

**Comment by kpedro88** [comment]
@suchandradutta please squash the commits in this PR to clean up the history (no need to preserve an accidental removal of code)

File: SimTracker/SiPhase2Digitizer/plugins/PSPDigitizerAlgorithm.cc [hunk 7]
```diff
@@ -68,3 +67,21 @@ void PSPDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterat
     ++simHitGlobalIndex;
   }
 }
+//
+// -- Select the Hit for Digitization
+//
+bool PSPDigitizerAlgorithm::select_hit(const PSimHit& hit, double tCorr, double& sigScale) {
+  bool result = false;
+  sigScale = 1.0;
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/PSPDigitizerAlgorithm.cc:75]: what's the purpose of this, since the variable is already set to 1 above (and then seemingly never used anywhere)?

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 8]
```diff
@@ -532,10 +532,11 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(
     }
   }
   // Fill the global map with all hit pixels from this event
+  float corr_time = hit.tof() - pixdet->surface().toGlobal(hit.localPosition()).mag() / 30.;
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc:535]: magic number 30 should be named constant

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 9]
```diff
@@ -976,7 +983,7 @@ int Phase2TrackerDigitizerAlgorithm::convertSignalToAdc(uint32_t detID, float si
         temp_signal = std::floor((temp_signal - kink_point) / (pow(2, dualslope_param - 1))) + kink_point;
     }
     signal_in_adc = std::min(temp_signal, theAdcFullScale_);
-    LogTrace("Phase2TrackerDigitizerAlgorithm")
+    LogInfo("Phase2TrackerDigitizerAlgorithm")
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc:986]: does this need to be available in production jobs? if not, please revert to LogTrace

File: SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc [hunk 10]
```diff
@@ -49,7 +49,10 @@ PixelDigitizerAlgorithm::PixelDigitizerAlgorithm(const edm::ParameterSet& conf)
               .getParameter<double>("Odd_column_interchannelCoupling_next_column")),
       even_column_interchannelCoupling_next_column_(
           conf.getParameter<ParameterSet>("PixelDigitizerAlgorithm")
-              .getParameter<double>("Even_column_interchannelCoupling_next_column")) {
+              .getParameter<double>("Even_column_interchannelCoupling_next_column")),
+      apply_timewalk_(conf.getParameter<ParameterSet>("PixelDigitizerAlgorithm").getParameter<bool>("ApplyTimeWalk")),
+      timewalk_model_(
+          conf.getParameter<ParameterSet>("PixelDigitizerAlgorithm").getParameter<edm::ParameterSet>("TimewalkModel")) {
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc:55]: try to use consistent capitalization in python parameters (Walk vs walk)

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc (hunk 8 above)
Comment (kpedro88) [reply SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc:535]: (in all the other files, also)

File: SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc [hunk 11]
```diff
@@ -203,3 +211,62 @@ void PixelDigitizerAlgorithm::add_cross_talk(const Phase2TrackerGeomDetUnit* pix
     }
   }
 }
+
+PixelDigitizerAlgorithm::TimewalkCurve::TimewalkCurve(const edm::ParameterSet& pset)
+    : x_(pset.getParameter<std::vector<double>>("charge")), y_(pset.getParameter<std::vector<double>>("delay")) {
+  if (x_.size() != y_.size())
+    throw cms::Exception("Configuration") << "Timewalk model error: Invalid curve.";
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc:218]: could make this message a bit more informative, like the one below: "the number of charge values does not match the number of delay values"

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 12]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+//
+bool SSDigitizerAlgorithm::select_hit(const PSimHit& hit, double tCorr, double& sigScale) {
+  bool result = false;
+  if (hitDetectionMode_ == SSDigitizerAlgorithm::SampledMode)
+    result = select_hit_sampledMode(hit, tCorr, sigScale);
+  else if (hitDetectionMode_ == SSDigitizerAlgorithm::LatchedMode)
+    result = select_hit_latchedMode(hit, tCorr, sigScale);
+  else {
+    double toa = hit.tof() - tCorr;
+    if (toa > theTofLowerCut_ && toa < theTofUpperCut_)
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:87]: simplify: `result = (toa > theTofLowerCut_ && toa < theTofUpperCut_);`

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 13]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+bool SSDigitizerAlgorithm::select_hit_sampledMode(const PSimHit& hit, double tCorr, double& sigScale) {
+  double toa = hit.tof() - tCorr;
+  double sampling_time = bx_time;
+
+  DetId det_id = DetId(hit.detUnitId());
+  float theThresholdInE =
+      (det_id.subdetId() == StripSubdetector::TOB) ? theThresholdInE_Barrel_ : theThresholdInE_Endcap_;
+
+  sigScale = getSignalScale(sampling_time - toa);
+  if (sigScale * hit.energyLoss() / GeVperElectron_ > theThresholdInE)
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:104]: simplify: `return (sigScale * hit.energyLoss() / GeVperElectron_ > theThresholdInE);`

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 14]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+  DetId det_id = DetId(hit.detUnitId());
+  float theThresholdInE =
+      (det_id.subdetId() == StripSubdetector::TOB) ? theThresholdInE_Barrel_ : theThresholdInE_Endcap_;
+
+  bool lastPulse = true;
+  bool aboveThr = false;
+  for (float i = deadTime_; i <= bx_time; i++) {
+    sigScale = getSignalScale(sampling_time - toa + i);
+
+    if (sigScale * hit.energyLoss() / GeVperElectron_ > theThresholdInE)
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:127]: simplify: `aboveThr = (sigScale * hit.energyLoss() / GeVperElectron_ > theThresholdInE);`

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 15]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+  return false;
+}
+
+double SSDigitizerAlgorithm::nFactorial(int n) { return TMath::Gamma(n + 1); }
+double SSDigitizerAlgorithm::aScalingConstant(int N, int i) {
+  return std::pow(-1, (double)i) * nFactorial(N) * nFactorial(N + 2) /
+         (nFactorial(N - i) * nFactorial(N + 2 - i) * nFactorial(i));
+}
+double SSDigitizerAlgorithm::cbc3PulsePolarExpansion(double x) {
+  if (pulseShapeParameters_.size() < 6)
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:145]: magic number 6 should be named constant

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 16]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+    return 0;
+
+  for (int i = 0; i < nTerms; i++) {
+    double angularTerm = 0;
+    double temporalTerm = 0;
+    double rTerm = std::pow(r, i) / (std::pow(tau, 2. * i) * nFactorial(i + 2));
+    for (int j = 0; j <= i; j++) {
+      double aij = nFactorial(i) * nFactorial(i + 2) / (nFactorial(i - j) * nFactorial(i + 2 - j) * nFactorial(j));
+      angularTerm += std::pow(std::cos(theta), (double)(i - j)) * std::pow(std::sin(theta), (double)j);
+      temporalTerm += std::pow(-1., (double)j) * aij * std::pow(xx, (double)(i - j)) * std::pow(tau, (double)j);
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:165]: isn't `std::pow(-1., (double)j) * aij` just `aScalingConstant(i,j)`?

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 17]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+    double fi = rTerm * angularTerm * temporalTerm;
+
+    fN += fi;
+  }
+  return fN;
+}
+double SSDigitizerAlgorithm::signalShape(double x) {
+  double xOffset = pulseShapeParameters_[0];
+  double tau = pulseShapeParameters_[1];
+  //  double r = pulseShapeParameters_[2];
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:176]: delete commented-out code

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 18]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+  //  double theta = pulseShapeParameters_[3];
+  //  double nTerms = pulseShapeParameters_[4];
+  double maxCharge = pulseShapeParameters_[5];
+
+  double xx = x - xOffset;
+  return maxCharge * (TMath::Exp(-xx / tau) * std::pow(xx / tau, 2.) * cbc3PulsePolarExpansion(x));
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:182]: use `std::exp()` rather than `TMath`

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 19]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+      aboveThr = false;
+    if (!lastPulse && aboveThr) {
+      return true;
+    }
+    lastPulse = aboveThr;
+  }
+  return false;
+}
+
+double SSDigitizerAlgorithm::nFactorial(int n) { return TMath::Gamma(n + 1); }
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:139]: use `std::tgamma()` rather than `TMath`

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 20]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+  //  double r = pulseShapeParameters_[2];
+  //  double theta = pulseShapeParameters_[3];
+  //  double nTerms = pulseShapeParameters_[4];
+  double maxCharge = pulseShapeParameters_[5];
+
+  double xx = x - xOffset;
+  return maxCharge * (TMath::Exp(-xx / tau) * std::pow(xx / tau, 2.) * cbc3PulsePolarExpansion(x));
+}
+void SSDigitizerAlgorithm::storeSignalShape() {
+  for (int i = 0; i < 1000; i++) {
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:185]: magic numbers (1000, 0.1) should be named constants

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 21]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+    float val = i * 0.1;
+
+    pulseShapeVec_.push_back(signalShape(val));
+  }
+}
+double SSDigitizerAlgorithm::getSignalScale(double xval) {
+  double res = 0.0;
+  int len = pulseShapeVec_.size();
+
+  if (xval * 10 >= len)
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:195]: magic number 10 should be named constant

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 22]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+    return res;
+  if (xval < 0.0)
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:197]: combine with above condition

File: SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc [hunk 23]
```diff
@@ -68,3 +73,146 @@ void SSDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iterato
...
+      res = pulseShapeVec_[i - 1];
+      break;
+    }
+  }
+  return res;
+}
+bool SSDigitizerAlgorithm::isAboveThreshold(const DigitizerUtility::SimHitInfo* const hisInfo,
+                                            float charge,
+                                            float thr) {
+  if (charge >= thr)
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc:214]: simplify: `return (charge >= thr);`

**Comment by civanch** [comment]
please test

**Comment by suchandradutta** [comment]
> The validation slides in the PR description seem to indicate there should be no changes from this PR. However, the comparisons show numerous changes. Are these expected?

I have checked histograms related to Phase2 Tracker Digis in the GUI for the work flow ([20034.0](https://cmsweb.cern.ch/dqm/dev/start?runnr=1;dataset%3D/RelVal_wf20034_0_pr/CMSSW_11_1_X-PRcmssw_29480-5766/DQMIO;referenceshow%3Dall;referencenorm=False;referenceobj1%3Dother::/RelVal_wf20034_0_base/CMSSW_11_1_X-PRcmssw_29480-5766/DQMIO::;sampletype%3Doffline_relval;workspace%3DEverything;), [20434.0](https://cmsweb.cern.ch/dqm/dev/start?runnr=1;dataset%3D/RelVal_wf20434_0_pr/CMSSW_11_1_X-PRcmssw_29480-5766/DQMIO;referenceshow%3Dall;referencenorm=False;referenceobj1%3Dother::/RelVal_wf20434_0_base/CMSSW_11_1_X-PRcmssw_29480-5766/DQMIO::;sampletype%3Doffline_relval;workspace%3DEverything;), [21234.0](https://cmsweb.cern.ch/dqm/dev/start?runnr=1;dataset%3D/RelVal_wf21234_0_pr/CMSSW_11_1_X-PRcmssw_29480-5766/DQMIO;referenceshow%3Dall;referencenorm=False;referenceobj1%3Dother::/RelVal_wf21234_0_base/CMSSW_11_1_X-PRcmssw_29480-5766/DQMIO::;sampletype%3Doffline_relval;workspace%3DEverything;), [23234.0](https://cmsweb.cern.ch/dqm/dev/start?runnr=1;dataset%3D/RelVal_wf23234_0_pr/CMSSW_11_1_X-PRcmssw_29480-5766/DQMIO;referenceshow%3Dall;referencenorm=False;referenceobj1%3Dother::/RelVal_wf23234_0_base/CMSSW_11_1_X-PRcmssw_29480-5766/DQMIO::;sampletype%3Doffline_relval;workspace%3DEverything;)) and found some folders are missing. But the existing histograms completely agree with reference. 

I myself have executed the workflow 21234.0 and age get all the folders as expected. 

I believe there is no issue with this PR and the fact that some folders are missing in the GUI must be for some other issue not related to this PR.

**Comment by mmusich** [comment]
> and found some folders are missing.

@suchandradutta the DQM by bin bin comparison shows only the plots that are detected as "changed", so you won't find all the folders but only the ones in which something has changed.

**Comment by civanch** [comment]
please test

**Comment by suchandradutta** [comment]
I have checked the histograms where differences are reported. They are not statistically significant.In other words the changes are not unexpected.

**Comment by kpedro88** [comment]
But why do they occur at all? Does this PR change the sequence of random number calls, or modify floating point arithmetic, or...?

**Comment by suchandradutta** [comment]
> But why do they occur at all? Does this PR change the sequence of random number calls, or modify floating point arithmetic, or...?

Yes, a few integer and float literals were defined as constant. For example,  we are now using 

constexpr double c_cm_ns = CLHEP::c_light * CLHEP::ns / CLHEP::cm;
constexpr double c_inv = 1.0 / c_cm_ns;
.....
float corr_time = hit.tof() - pixdet->surface().toGlobal(hit.localPosition()).mag() * c_inv;

instead of 

float corr_time = hit.tof() - pixdet->surface().toGlobal(hit.localPosition()).mag() / 30.;   

in SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc and in other places.

**Comment by kpedro88** [comment]
Ah, great. Can you update the PR description so this change is clearly described?

**Comment by kpedro88** [comment]
+upgrade
minor changes caused by improved accuracy in unit conversions

**Comment by civanch** [comment]
+1

**Comment by silviodonato** [comment]
please test workflow 23234.1001, 23434.1001

**Comment by silviodonato** [comment]
+1
FYI @fwyzard @trtomei I checked that the output of  23434.1001, 23234.1001 does not changes dramatically.

**Comment by fwyzard** [comment]
Thanks for the check and heads up.

I understand that this PR changes the production of new DIGIs during the simulation step - but has no effect on the unpacking of existing RAW or reading existing DIGIs.

**Comment by suchandradutta** [comment]
> Thanks for the check and heads up.
> 
> I understand that this PR changes the production of new DIGIs during the simulation step - but has no effect on the unpacking of existing RAW or reading existing DIGIs.

@fwyzard this PR will not change the Digi collection. The option of time dependent signal shape has been included but at present it is not switched off. Hence the Digis should be identical apart from the minor changes due to improved accuracy in a few constants!
